### PR TITLE
Make Channel implement IChannel

### PR DESCRIPTION
**Pull request by davidfowl**
- Remove IReadableChannel and IWritableChannel from Channel impl
- Update all the places assumed a direct cast was possible

Thoughts?

/cc @mgravell @benaadams @halter73 @pakrym

### CODE REVIEW

File: src/Channels.Networking.Libuv/UvTcpConnection.cs [hunk 1]
```diff
@@ -37,24 +37,24 @@ public UvTcpConnection(UvThread thread, UvTcpHandle handle)
...
+        public IReadableChannel Input => _input.Input;
 
         private async Task ProcessWrites()
         {
             try
             {
                 while (true)
                 {
-                    var buffer = await _output.ReadAsync();
+                    var buffer = await _output.Input.ReadAsync();
```
Comment (benaadams) [inline src/Channels.Networking.Libuv/UvTcpConnection.cs]: Shouldn't this be using a property? or have some renaming? The 180° in naming in the same statement jars

Comment (davidfowl) [reply src/Channels.Networking.Libuv/UvTcpConnection.cs]: Yea it needs a better name

**Comment by mgravell** [comment]
I'm conflicted about this, partly because I wonder if I proposed the wrong name in the original suggestion.  I wonder if it should have been `IConnection` or `ITransport`. The expectation there - in my imagination - is the many cases where something has **independent** input/output streams. However, `Channel` is slightly different - here the input _is_ the output - a single channel. This is not the same scenario as the one I was really talking about previously, which allows abstraction over transports. And indeed, this would allow someone to pass a `Channel` _to_ one of those abstractions, which ... would have very odd and unusual effects - it would be talking to itself in a loop, almost certainly not what was intended.

I wonder if we've crossed intents and spoken past each-other.

**Comment by davidfowl** [comment]
> I wonder if we've crossed intents and spoken past each-other.

Naw, we didn't it just feels like there's an opportunity for a single abstraction here. Kinda like `MemoryStream`,`NetworkStream`, `FooStream`. The split between input and output is also nice from a code POV. The default channel implementation doesn't need to be both the reader and the writer. Properties can hang off it. I always wanted to do that and this basically give me an easy way that unified with what could be the one true channel interface.

**Pull request closed without merging**